Fix handling of stopped TUI applications on unix

nome opened the pull request:
# Description

Instead of handling a foreground process being stopped in any way, we were simply ignoring SIGTSTP (which the terminal usually sends to the foreground process group when Ctrl-Z is pressed), and propagating this to our foreground children. This works for most processes, but it generally fails for applications which put the terminal in raw mode[1] and implement their own suspension mechanism (typically TUI apps like helix[2], neovim[3], top[4] or amp[5]). In these cases, triggering suspension within the app results in the terminal getting blocked, since the application is waiting for a SIGCONT, while nushell is waiting for it to exit.

Fix this by unblocking SIGTSTP for our foreground children (neovim, helix and probably others send this to themselves while trying to suspend), and displaying the following message whenever one of them gets stopped:

    nushell currently does not support background jobs
    press any key to continue

Pressing any key will then send SIGCONT to the child's process group and resume waiting.

This fix is probably going to be superseded by a proper background job implementation (#247) at some point, but for now it's better than completely blocking the terminal.

[1] https://docs.rs/crossterm/latest/crossterm/terminal/index.html#raw-mode
[2] https://helix-editor.com/
[3] https://neovim.io/
[4] https://man7.org/linux/man-pages/man1/top.1.html
[5] https://amp.rs/

- fixes #1038
- fixes #7335
- fixes #10335

# User-Facing Changes

While any foreground process is running, Ctrl-Z is no longer ignored. Instead, the message described above is displayed, and nushell waits for a key to be pressed.

# Tests + Formatting

<!--
Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used` to check that you're using the standard code style
- `cargo test --workspace` to check that all tests pass (on Windows make sure to [enable developer mode](https://learn.microsoft.com/en-us/windows/apps/get-started/developer-mode-features-and-debugging))
- `cargo run -- -c "use toolkit.nu; toolkit test stdlib"` to run the tests for the standard library

> **Note**
> from `nushell` you can also use the `toolkit` as follows
> ```bash
> use toolkit.nu  # or use an `env_change` hook to activate it automatically
> toolkit check pr
> ```
-->

# After Submitting
<!-- If your PR had any user-facing changes, update [the documentation](https://github.com/nushell/nushell.github.io) after the PR is merged, if necessary. This will help us keep the docs up to date. -->

WindSoilder commented:
@IanManske Can you take a look at this when you get time?

IanManske reviewed (commented):
Thanks for the PR! I think this is a good thing to add before proper job control.

Would you mind moving `process/exit_status.rs` from `nu_protocol` into `nu_system`? This way, you should be able to use `nix::sys::wait::waitpid` if you return the custom `ExitStatus` from this function. Thanks!

nome commented:
I didn't find as much time for this during the weekend as I had hoped; but in general it looks doable.

Since this will involve reverting some of the previous changes, would you prefer to have the modifications on top of the current branch or force-pushed with a cleaned up history?

IanManske commented:
> would you prefer to have the modifications on top of the current branch or force-pushed with a cleaned up history?

Either is fine, whichever works best for you. Also no rush :)

nome commented:
Moving `nu_protocol::process::exit_status` was conflicting with your changes in #13515, so I did that on top of the current main branch and rebased the actual fix. While I was at it, I also cleaned up handling of the various `WaitStatus` variants.

fdncred commented:
sorry, there are conflicts now. would you mind fixing those?

nome commented:
> sorry, there are conflicts now

I've rebased the changes, should be ready to merge now

fdncred commented:
Thanks. Let's try this out.

Deebster commented:
This is working for me, thank you @nome

nome commented:
> This is working for me, thank you @nome

You're welcome. It was my main pain point with nushell; since I got this fixed, I'm using nu as daily driver. :)
